README.md
# Selector – Kirby Fileselect Field

This additional panel field for [Kirby 2](http://getkirby.com) allows you to use an intuitive alternative file selection field in your blueprints.

**Authors**: [digital storytelling pioneers](https://github.com/storypioneers) feat. [Jonas Doebertin](https://github.com/JonasDoebertin)

**License**: [GNU GPL v3.0](http://opensource.org/licenses/GPL-3.0)

![Screenshot](https://raw.githubusercontent.com/storypioneers/kirby-selector/master/screenshot.png)

## Installation

Copy or link the `selector` directory to `site/plugins/` **or** use the [Kirby CLI](https://github.com/getkirby/cli) `kirby plugin:install storypioneers/kirby-selector`.


## Usage

### Within your blueprints

Post installation simply add `selector` fields to your blueprints and set some options (where applicable).

```yaml
fields:
	postimage:
		label: Main Post Image
		type:  selector
		mode:  single
		types:
			- image
```
```yaml
fields:
	attachments:
		label: Post Attachments
		type:  selector
		mode:  multiple
		types:
			- all
```

### Within your templates / controllers / models

As per design, the selector field stores the *filenames* of the selected files, only. If you need access to the files full path or to other properties / functions of the file object, you must convert the filename into a full file object first.

**Single Mode**

When you're using the Selector field in Single Mode, gaining access to the full file object is quite easy. Just replace `yourselectorfield` with the name of your Selector-based field.

```php
	// Convert the filename to a full file object
	$file = $page->yourselectorfield()->toFile();

	// Use the file object
	echo $file->url();
```

**Multiple Mode**

In multiple mode, the Selector field stores a comma-separated list of filenames, based on how many files you selected. To convert this list into a fully-featured file collection (just like `$page->files()`), you need a bit more code.

```php

	// Transform the comma-separated list of filenames into a file collection
	$filenames = $page->yourselectorfield()->split(',');
	if(count($filenames) < 2) $filenames = array_pad($filenames, 2, '');
	$files = call_user_func_array(array($page->files(), 'find'), $filenames);

	// Use the file collection
	foreach($files as $file)
	{
		echo $file->url();
	}

```

## Options

The field offers some options that can be set on a per field basis directly from your blueprints.

### mode

Define a mode the field will work in. Possible values are `single` and `multiple`.

* **single**: With the `single` mode, the fields checkboxes will work like a set of radio buttons, letting you select a single file only. This is useful if you want a way to specify a posts main image.

* **multiple**: This option allows you to select multiple files in a single file selector field. The selected files will be stored as a comma separated list.

### types

Define the files types the file selector field shows. Possible values are `all`, `image`, `video`, `audio`, `document`, `archive`, `code`, `unknown`. You may specify as many of these types as you like.

```yaml
fields:
	attachments:
		label: Post Attachments
		type:  selector
		mode:  multiple
		types:
			- image
			- video
			- audio
			- document
```

### sort

Files will be shown sorted by their filename in ascending order (a-z). However, this option let's you change the default sort behavior. You can sort files by filesize, dimensions, type and many more. Some of the countless possible values are `sort`, `filename`, `size`, `width`, `height`, `type`, `modified`, `ratio`. You can also sort by any of your custom [file fields](http://getkirby.com/docs/panel/blueprints/file-settings#file-fields). The value `sort` will make sure the files are presented in the exact order you specified in the panels file section via [drag and drop](http://getkirby.com/docs/panel/blueprints/file-settings#sortable-files).

```yaml
fields:
	featured:
		label: Featured Image
		type:  selector
		mode:  single
		sort:  size
		types:
			- image
```

### flip

This options allows you to reverse the sort order you specified with the `sort` option. You may set this to `true` or `false`.

### autoselect

This options allows you to tell the Selector to auto select the first, last or all files of the list, if no other file is selected, yet. Possible values are `none` (default), `first`, `last` and `all`.

```yaml
fields:
	featured:
		label:      Featured Image
		type:       selector
		mode:       single
		sort:       filename
		autoselect: first
		types:
			- image
```

### autoselectadd

This options allows you to tell the Selector to auto select a file if the upload is initiated by clicking on the field-related _Add_ button.

```yaml
fields:
	featured:
		label:         Featured Image
		type:          selector
		mode:          single
		sort:          filename
		autoselectadd: true
		types:
			- image
```

### filter

This options allows you to set a filename filter. This can be either a simple string or a fully featured regular expression. Only files with filenames matching the filter string or regular expression will be shown in the Selector field. You may set this to any string like `background`, `.min.js` or `large` or a regular expression like `/\.((png)|(jpe?g))/i`.

```yaml
fields:
	featured:
		label:  Page Background Image
		type:   selector
		mode:   single
		filter: background
		types:
			- image
```
Showing only image files with the term *background* in their filename.

```yaml
fields:
	featured:
		label:  Page Background Image
		type:   selector
		mode:   single
		filter: /\.((png)|(jpe?g))/i
```
Using a regular expression to show only *.jpg*, *.jpeg* and *.png* files.

### size

The `size` option lets you limit the height of the selector field and makes it scrollable. Only the specified number of files will be shown initially. This can be set either to a *number* or to `auto` (the Selector field will adapt to the height of the complete list of files).

```yaml
fields:
	featured:
		label: Page Background Image
		type:  selector
		mode:  single
		size:  4
```


CHANGELOG.md
# Changelog

## Version 1.6.0

**Features:**

* Add support for auto selection after uploding a file using the field specific _Add_ button

## Version 1.5.2

**Features:**

* Kirby CLI support


## Version 1.5.1

**Bugfixes:**

* More compatible way of getting the current panel language

## Version 1.5.0

**Features:**

* Full Kirby 2.2 support
* Context menus (Kirby 2.2+)

**Changes:**

* Better list item layout — now fully responsive
* Better "No files found" message

**Bugfixes:**

* Don't show empty (0x0) dimensions for svg images

## Version 1.4.2

**Bugfixes:**

* Fixed PHP 5.3 compatibility
* Show "No matching files" message even if a max size is set

## Version 1.4.1

**Changes:**

* Improved styling for selected items

**Additional:**

* Added a list of all contributors

## Version 1.4.0

**Features:**

* Allow regular expressions for the filter option
* Add new size setting to limit the field height

**Changes:**

* Minor code cleanup/reformatting (PSR-2)

## Version 1.3.0

**Features:**

* Filter by filename
* Support for Kirby 2.1+ site files
* Better documentation

**Bugfixes:**

* Don't break on fields without labels

## Version 1.2.0

**Features:**

* Add translation support
* Add `autoselect` option

## Version 1.1.0

**Features:**

* Introduced sorting options (see #8)

**Bugfixes:**

* Allow unselecting in single mode (see #5)

## Version 1.0.2

**Bugfixes:**

* Fixed a PHP Warning

## Version 1.0.1

**Bugfixes:**

* Added support for PHP 5.3

## Version 1.0.0

* Initial public release


field/selector.php
<?php

class SelectorField extends BaseField
{
    /**
     * Base directory for language files.
     *
     * @var string
     * @since 1.2.0
     */
    const LANG_DIR = 'languages';

    /**
     * Define frontend assets.
     *
     * @var array
     * @since 1.0.0
     */
    public static $assets = array(
        'css' => array(
            'selector.css',
        ),
        'js' => array(
            'selector.js',
        ),
    );

    /**
     * Select mode (single/multiple).
     *
     * @var string
     * @since 1.0.0
     */
    protected $mode = 'single';

    /**
     * Sort mode.
     *
     * @var string
     * @since 1.1.0
     */
    protected $sort = 'filename';

    /**
     * Flip sort order.
     *
     * @var string
     * @since 1.1.0
     */
    protected $flip = false;

    /**
     * Covered file types.
     *
     * @var array
     * @since 1.0.0
     */
    protected $types = array('all');

    /**
     * Autoselect a file.
     *
     * @var string
     * @since 1.2.0
     */
    protected $autoselect = 'none';

    /**
     * Autoselect a file after use the field specific add button.
     *
     * @var string
     * @since 1.6.0
     */
    protected $autoselectadd = 'none';

    /**
     * Filename filter.
     *
     * @var bool|string
     * @since 1.3.0
     */
    protected $filter = false;

    /**
     * Selector size (number of visible items).
     *
     * @var string|int
     * @since 1.4.0
     */
    protected $size = 'auto';

    /**
     * Selector size (number of visible items).
     *
     * @var string|int
     * @since 1.4.0
     */
    protected $editable = true;

    /**
     * Option default values.
     *
     * @var array
     * @since 1.0.0
     */
    protected $defaultValues = array(
        'mode'    => 'single',
        'options' => 'all',
        'size'    => 'auto',
    );

    /**
     * Valid option values.
     *
     * @var array
     * @since 1.0.0
     */
    protected $validValues = array(
        'mode'  => array(
            'single',
            'multiple',
        ),
        'types' => array(
            'all',
            'image',
            'video',
            'audio',
            'document',
            'archive',
            'code',
            'unknown',
        ),
        'autoselect' => array(
            'none',
            'first',
            'last',
            'all',
        ),
        'autoselectadd' => array(
            'none',
            'true',
        ),
    );

    /**
     * Field setup.
     *
     * @since 1.2.0
     *
     * @return \SelectorField
     */
    public function __construct()
    {
        // Build translation file path
        $baseDir = __DIR__ . DS . self::LANG_DIR . DS;

        // Get panel language
        if (version_compare(panel()->version(), '2.2', '>=')) {
            $lang = panel()->translation()->code();
        } else {
            $lang = panel()->language();
        }

        // Load language files
        if (file_exists($baseDir . $lang . '.php')) {
            require $baseDir . $lang . '.php';
        } else {
            require $baseDir . 'en.php';
        }
    }

    /**
     * Magic setter.
     *
     * Set a fields property and apply default value if required.
     *
     * @since 1.0.0
     *
     * @param string $option
     * @param mixed  $value
     */
    public function __set($option, $value)
    {
        /* Set given value */
        $this->$option = $value;

        /* Check if value is valid */
        switch ($option) {
            case 'mode':
                if (!in_array($value, $this->validValues['mode'])) {
                    $this->mode = $this->defaultValues['mode'];
                }
                break;

            case 'types':
                if (!is_array($value) or empty($value)) {
                    $this->types = array('all');
                }
                break;

            case 'sort':
                if (!is_string($value) or empty($value)) {
                    $this->sort = 'filename';
                }
                break;

            case 'flip':
                if (!is_bool($value)) {
                    $this->flip = false;
                }
                break;

            case 'autoselect':
                if (!in_array($value, $this->validValues['autoselect'])) {
                    $this->autoselect = 'none';
                }
                break;

            case 'autoselectadd':
                if (!in_array($value, $this->validValues['autoselectadd'])) {
                    $this->autoselectadd = 'none';
                }
                break;

            case 'filter':
                if (!is_string($value) or empty($value)) {
                    $this->filter = false;
                }
                break;

            case 'size':
                if (!is_numeric($value)) {
                    $this->size = $this->defaultValues['size'];
                }
                break;

            case 'editable':
                if (!is_bool($value)) {
                    $this->size = $this->defaultValues['editable'];
                }
                break;
        }
    }

    /**
     * Generate label markup.
     *
     * @since 1.0.0
     *
     * @return string
     */
    public function label()
    {
        /* Action button */
        $action = new Brick('a');
        $action->addClass('file-add-button label-option');
        $action->html('<i class="icon icon-left fa fa-plus-circle"></i>' . l('pages.show.files.add'));

        /**
         * FIX: With Kirby 2.2 the structure of the "Add file" actions changed.
         * Let's make sure we handle both the old >2.1 and the new 2.2+ ways.
         *
         * @since 1.5.0
         */
        if (version_compare(Kirby::version(), '2.2', '>=')) {
            $action->attr('href', '#upload');
            $action->data('upload', 'true');
        } else {
            $action->attr('href', purl($this->page(), 'upload'));
        }

        /* Label */
        $label = parent::label();

        /**
         * FIX: Fields don't have to have a label assigned.
         * With this, we deal with missing label information.
         *
         * @since 1.3.0
         */
        if (!is_null($label)) {
            $label->addClass('figure-label');

            // add upload load link if field is editable
            if($this->editable) {
                $label->append($action);
            }

            return $label;
        }

        return;
    }

    /**
     * Generate field content markup.
     *
     * @since 1.0.0
     *
     * @return string
     */
    public function content()
    {
        $wrapper = new Brick('div');
        $wrapper->addClass('selector');
        $wrapper->data(array(
            'field'         => 'selector',
            'name'          => $this->name(),
            'page'          => $this->page(),
            'mode'          => $this->mode,
            'autoselect'    => $this->autoselect(),
            'autoselectadd' => $this->autoselectadd(),
            'size'          => $this->size,
            'editable'      => $this->editable,
        ));
        $wrapper->html(tpl::load(__DIR__ . DS . 'template.php', array('field' => $this)));

        return $wrapper;
    }

    /**
     * Return the current value.
     *
     * @since 1.0.0
     *
     * @return array
     */
    public function value()
    {
        if (is_string($this->value)) {
            $this->value = str::split($this->value, ',', 1);
        }

        return $this->value;
    }

    /**
     * Get files based on types option.
     *
     * @since 1.0.0
     *
     * @return \Files
     */
    public function files()
    {
        /**
         * FIX: When used in site options, we don't have a `$this->page`
         * property we can use to access the pages files.
         *
         * (1) If we have page property, we'll use that to fetch the files.
         * (2) If we don't have a page property we're on the site options page.
         *     (2.1) If we're using Kirby 2.1+ we can use the new `site()->files()`
         *           method to get access to the new global site files to use them.
         *     (2.2) If we are using a lower version, global site files don't
         *           exist. We'll return an empty collection object instead.
         *
         * @since 1.3.0
         */
        if (!is_null($this->page)) { /* (1) */
            $files = $this->page->files(); /* (1) */
        } else { /* (2) */
            if (version_compare(Kirby::version(), '2.1', '>=')) { /* (2.1) */
                $files = site()->files(); /* (2.1) */
            } else {
                return new Collection(); /* (2.2) */
            }
        }

        /**
         * FIX: Create a new reference to $this to overcome the unavailability
         * of $this within closures in PHP < 5.4.0 by passing this new reference
         * with the "use" language construct.
         *
         * @since 1.0.1
         */
        $field = &$this;

        $files = $files->sortBy($this->sort, ($this->flip) ? 'desc' : 'asc')
            ->filter(function ($file) use ($field) {
                return $field->includeAllFiles() or in_array($file->type(), $field->types());
            });

        /**
         * Filter files using a regular expression.
         *
         * @since 1.4.0
         */
        if ($this->isRegExp($this->filter)) {
            $files = $files->filter(function ($file) use ($field) {
                return (preg_match($this->filter, $file->filename()) === 1);
            });
        }

        /**
         * Filter files by filename if a filter has been set.
         *
         * @since 1.3.0
         */
        elseif ($this->filter) {
            $files = $files->filterBy('filename', '*=', $this->filter);
        }

        return $files;
    }

    /**
     * Generate file slug.
     *
     * @since 1.0.0
     *
     * @param  \File  $file
     * @return string
     */
    public function itemId($file)
    {
        return $this->name() . '-' . str::slug($file->filename());
    }

    /**
     * Check if a file is present in the current value.
     *
     * @since 1.0.0
     *
     * @param  \File $file
     * @return bool
     */
    public function isInValue($file)
    {
        return in_array($file->filename(), $this->value());
    }

    /**
     * Check if the types array includes "all".
     *
     * @since 1.0.0
     *
     * @return bool
     */
    public function includeAllFiles()
    {
        return in_array('all', $this->types);
    }

    /**
     * Check if a string is a valid regular expression.
     *
     * @since  1.4.0
     *
     * @param  string $string
     * @return bool
     */
    public function isRegExp($string)
    {
        return !(@preg_match($string, null) === false);
    }

    /**
     * Return type filter array.
     *
     * FIX: This function is required since we create a new variable containing
     * a reference to `$this` to overcome the unavailability of `$this within
     * closures in PHP < 5.4.0 by passing this new reference with the "use"
     * language construct in the `files()` function.
     *
     * @since  1.4.2
     * @return array
     */
    public function types()
    {
        return $this->types;
    }
}
